db/cache_db.go

```diff
@@ -0,0 +1,157 @@
+package db
+
+import (
+	"fmt"
+	"sort"
+	"sync"
+)
+
+// If value is nil but deleted is false,
+// it means the parent doesn't have the key.
+// (No need to delete upon Commit())
+type cDBValue struct {
+	value   []byte
+	deleted bool
+}
+
+// CacheDB wraps an in-memory cache around an underlying DB.
+type CacheDB struct {
+	mtx    sync.Mutex
+	cache  map[string]cDBValue
+	parent DB
+}
+
+// NOTE: Needed by MultiStore.CacheWrap().
+var _ atomicSetDeleter = (*CacheDB)(nil)
+
+func NewCacheDB(parent DB) *CacheDB {
+	db := &CacheDB{
+		cache:  make(map[string]cDBValue),
+		parent: parent,
+	}
+	return db
+}
+
+func (db *CacheDB) Get(key []byte) []byte {
+	db.mtx.Lock()
+	defer db.mtx.Unlock()
+
+	value, ok := db.cache[string(key)]
+	if !ok {
```

Comment:
One other check we might need, I had mentioned in the previous review but in later comments: we perhaps want to check that value.deleted is not set lest that's a discarded value in the cache i.e
```go
if !ok || value.deleted {
}
```
What do y'all think?

Reply:
not value.deleted... if value.deleted in the cache, we don't want to query the base store.

if value.deleted, value.value is nil btw.